nx/lib/nx/defn/kernel.ex

```diff
@@ -1075,6 +1105,197 @@ defmodule Nx.Defn.Kernel do
+
+  The `defn` above defines two hooks, one is called with the
+  value of `a + b` and another with `a * b`. Once you invoke
+  the function above, you should see this printed:
+
+      Hooks.add_and_mult(2, 3)
+      {:add, #Nx.Tensor<
+         s64
+         5
+      >}
+      {:mult, #Nx.Tensor<
+         s64
+         6
+      >}
+
+  In other words, the `hook` function accepts a tensor
+  expression as argument and it will invoke a custom
+  function with a tensor value at runtime. `hook` returns
+  the result of the given expression. The expression can
+  be any tensor or a `Nx.Container`.
+
+  Note **you must return the result of the `hook` call**.
+  For example, the code below won't inspect the `:add`
+  tuple, because the hook is not returned from `defn`:
+
+      defn add_and_mult(a, b) do
+        _add = hook(a + b, fn tensor -> IO.inspect({:add, tensor}) end)
+        mult = hook(a * b, fn tensor -> IO.inspect({:mult, tensor}) end)
+        mult
+      end
+
+  We will learn how to hook into a value that is not part
+  of the result in the "Hooks and tokens" section.
+
+  ## Named hooks
+
+  It is possible to give names to the hooks. This allows them
+  to be defined or overridden by calling `Nx.Defn.jit/3` or
+  `Nx.Defn.stream/3`. Let's see an example:
+
+      defmodule Hooks do
+        import Nx.Defn
+
+        defn add_and_mult(a, b) do
+          add = hook(a + b, :hooks_add)
+          mult = hook(a * b, :hooks_mult)
+          {add, mult}
+        end
+      end
+
+  Now you can pass the hook as argument as follows:
+      
+      hooks = %{
+        hooks_add: fn tensor ->
+          IO.inspect {:add, tensor}
+        end
+      }
+
+      args = [Nx.tensor(2), Nx.tensor(3)]
+      Nx.Defn.jit(&Hooks.add_and_mult/2, args, hooks: hooks)
+
+  > **Important!** We recommend to prefix your hook names
+  > by the name of your library to avoid conflicts.
+
+  If a named hook is not given, compilers can optimize
+  that away and not transfer the tensor from the device
+  in the first place.
+
+  You can also mix named hooks with callbacks:
+
+      defn add_and_mult(a, b) do
+        add = hook(a + b, :hooks_add, fn tensor -> IO.inspect({:add, tensor}) end)
+        mult = hook(a * b, :hooks_mult, fn tensor -> IO.inspect({:mult, tensor}) end)
+        {add, mult}
+      end
+
+  If a hook with the same name is given to `Nx.Defn.jit/3`
+  or `Nx.Defn.stream/3`, then it will override the default
+  callback.
+
+  ## Hooks and tokens
+
+  So far, we have always return the result of the `hook`
+  call. However, what happens if the values we want to
+  hook are not part of the return value, such as below?
+
+      defn add_and_mult(a, b) do
+        _add = hook(a + b, :hooks_add, &IO.inspect({:add, &1}))
+        mult = hook(a * b, :hooks_mult, &IO.inspect({:mult, &1}))
+        mult
+      end
+
+  In such cases, you must use tokens. Tokens are used to
+  create an ordering over hooks, ensuring hooks execute
+  in a certain order:
```

Comment:
```suggestion
  in a certain sequence:
```
WDYT, to avoid repeating "order"?